File: A_ChainAnimationObjcDemo/ViewController.m
//
//  ViewController.m
//  A_ChainAnimationObjcDemo
//
//  Created by Animax Deng on 3/7/17.
//  Copyright © 2017 Animx. All rights reserved.
//

#import "ViewController.h"
#import "A_ChainAnimation.h"

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ViewController {
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
}

#pragma mark - implement UITableViewDelegate and UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 9;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    
    __weak UIView *box = [cell viewWithTag:1];
    __weak UILabel *titleLabel = (UILabel *)[cell viewWithTag:10];
    
    [[box layer] setTransform:CATransform3DIdentity];
    [[box layer] setBackgroundColor:[UIColor blackColor].CGColor];
    
    [titleLabel setText:@""];
    [box setHidden:NO];
    
    switch (indexPath.row) {
        case 0:
            if (box) {
                [titleLabel setText:@"UIView block animation chain example"];
                
                [[[box addAnimateWithDuration:1.0 aniamtion:^{
                    [box setAlpha:0.0];
                }] addAnimateWithDuration:1.0 aniamtion:^{
                    [box setAlpha:1.0];
                }] play];
            }
            break;
        case 1:
            if (box) {
                [titleLabel setText:@"Effection animation chain example"];
                
                [[[box addAnimateWithEffect:A_AnimationEffectType_pulse type:A_AnimationType_easeInBack duration:1.0]
                  addAnimateWithEffect:A_AnimationEffectType_squeeze type:A_AnimationType_bigSpring duration:1.0]
                 play];
            }
            break;
        case 2:
            if (box) {
                [titleLabel setText:@"UIView block animation and Effection animation chain mix example"];
                
                [[[[[box addAnimateWithDuration:1.0 aniamtion:^{
                    [box setAlpha:0.3];
                }] addAnimateWithEffect:A_AnimationEffectType_pulse type:A_AnimationType_easeInBack duration:1.0]
                   addAnimateWithDuration:1.0 aniamtion:^{
                       [box setAlpha:1.0];
                   }] addAnimateWithEffect:A_AnimationEffectType_squeeze type:A_AnimationType_easeOutElastic duration:1.0]
                 play];
            }
            break;
        case 3:
            if (box) {
                [titleLabel setText:@"Sync effection animation chain combin example"];
                
                [[[[box syncAnimate] addAnimateWithEffect:A_AnimationEffectType_wobble type:A_AnimationType_easeInBack duration:1.0]
                  addAnimateWithEffect:A_AnimationEffectType_mirror_zoomOut type:A_AnimationType_easeInExpo duration:1.5] play];
            }
            break;
        case 4:
            if (box) {
                [titleLabel setText:@"CALayer animation chain example"];
                
                [[[[[box syncAnimate]
                   setPositionX:20 AnimtionType:A_AnimationType_spring Duraion:2.0]
                  setSize:CGSizeMake(5, 5) AnimtionType:A_AnimationType_bigLongSpring Duraion:3.0].then
                 setCornerRadius:10 AnimtionType:A_AnimationType_noEffect]
                 play];
            }
            break;
        case 5:
            if (box) {
                [titleLabel setText:@"Custom oblique animation chain example"];

                [[[[[[[box animateWait:0.5] setLeftOblique:1.0 AnimtionType:A_AnimationType_easeInBack Duraion:1.0]
                     setTopOblique:1.0 AnimtionType:A_AnimationType_easeInQuad Duraion:1.0]
                    setRightOblique:1.0 AnimtionType:A_AnimationType_easeInCirc Duraion:1.0]
                   setBottomOblique:1.0 AnimtionType:A_AnimationType_easeInOutSine Duraion:1.0]
                  setRecoverOblique:A_AnimationType_easeInOutBounce Duraion:1.0]
                 play];

            }
            break;
        case 6:
            if (box) {
                [titleLabel setText:@"Wait, block, and animation chain example"];
                
                [[[[[[[box animateWait:2.0] setScale:1.5 AnimtionType:A_AnimationType_bigSpring Duraion:1.0] wait:1.0] block:^{
                    box.backgroundColor = [UIColor blueColor];
                }] wait:3.0]
                  setBackgroundColor:[UIColor redColor] AnimtionType:A_AnimationType_bigLongSpring Duraion:2.0] play];
            }
            break;
        case 7:
            if (box) {
                [titleLabel setText:@"One-time Effection Animation"];
                
                [box A_AnimationEffect:A_AnimationEffectType_press Repeat:2.5 Duration:1.0];
            }
            break;
        case 8:
            if (box) {
                [titleLabel setText:@"One-time CGLayer Animation"];
                [box.layer A_AnimationSetScaleX:1.5 AnimtionType:A_AnimationType_easeInOutBounce Duraion:1.0];
            }
            break;
        default:
            break;
    }
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 100;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    
}

@end

File: ObjectiveC_Version/A_ChainAnimationSwiftDemo/ViewController.swift
//
//  ViewController.swift
//  A_ChainAnimationSwiftDemo
//
//  Created by Animax Deng on 3/7/17.
//  Copyright © 2017 Animx. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var tableView: UITableView!
    override func viewDidLoad() {
        super.viewDidLoad()
    }

    // MARK: Implement UITableViewDelegate and UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 7
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell")
        
        if let box = cell?.viewWithTag(1), let titleLabel = cell?.viewWithTag(10) as? UILabel {
            
            titleLabel.text = ""
            box.isHidden = false
            
            switch indexPath.row {
            case 0:
                titleLabel.text = "UIView block animation chain example"
                box.addAnimate(withDuration: 1.0, aniamtion: { 
                    box.alpha = 0.0
                }).addAnimate(withDuration: 1.0, aniamtion: { 
                    box.alpha = 1.0
                }).play()
                
                break
            case 1:
                titleLabel.text = "Effection animation chain example"
                box.addAnimate(with: .pulse, type: .easeInBack, duration: 1.0)
                    .addAnimate(with: .squeeze, type: .bigSpring, duration: 1.0)
                    .play()
                break
            case 2:
                titleLabel.text = "UIView block animation and Effection animation chain mix example"
                
                box.addAnimate(withDuration: 1.0, aniamtion: { box.alpha = 0.3 })
                    .addAnimate(with: .pulse, type: .easeInBack, duration: 1.0)
                    .addAnimate(withDuration: 1.0, aniamtion: {
                        box.alpha = 1.0
                    })
                    .addAnimate(with: .squeeze, type: .easeOutElastic, duration: 1.0)
                    .play()
                
                break
            case 3:
                titleLabel.text = "Sync effection animation chain combin example"
                
                box.syncAnimate()
                    .addAnimate(with: .wobble, type: .easeInBack, duration: 1.0)
                    .addAnimate(with: .mirror_zoomOut, type: .easeInExpo, duration: 1.5)
                    .play()
                
                break
            case 4:
                titleLabel.text = "CALayer animation chain example"
                
                box.syncAnimate()
                    .setPositionX(20, animtionType: .spring, duraion: 2.0)
                    .setSize(CGSize(width: 5, height: 5), animtionType: .bigLongSpring , duraion: 3.0)
                    .then
                    .setCornerRadius(10, animtionType: .noEffect)
                    .play()
                
                break
            case 5:
                titleLabel.text = "Custom oblique animation chain example"
                
                box.animateWait(0.5).setLeftOblique(1.0, animtionType: .easeInBack, duraion: 1.0)
                    .setTopOblique(1.0, animtionType: .easeInQuad, duraion: 1.0)
                    .setRightOblique(1.0, animtionType: .easeInCirc, duraion: 1.0)
                    .setBottomOblique(1.0, animtionType: .easeInOutSine, duraion: 1.0)
                    .setRecoverOblique(.easeInOutBounce, duraion: 1.0)
                    .play()
                
                break
            case 6:
                titleLabel.text = "Wait, block, and animation chain example"
                
                box.animateWait(2.0)
                    .setScale(1.5, animtionType: .bigSpring, duraion: 1.0)
                    .wait(1.0)
                    .block({
                        box.backgroundColor = UIColor.blue
                    })
                    .wait(3.0)
                    .setBackgroundColor(UIColor.red, animtionType: .bigLongSpring, duraion: 2.0)
                    .play()
                
                break
            default:
                break
            }
        }
        
        return cell!
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100
    }
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath as IndexPath, animated: true)
    }
    
    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        
    }
    
}

